### PR TITLE
Simplifying configuration to use just environment variables

### PR DESCRIPTION
**Pull request by cschroed-usgs**
EDIT:
Right now in `master` the base image defines some environment variables. It then maps the environment variable's values to something that Spring can use in `application.yml` . However, Spring Boot isn't limited to reading config from `application.yml`; it supports reading configuration from environment variables too. This pull request proposes simplifying things by cutting out the `application.yml` middleman and tweaking the environment variables' names to match the names Spring expects.
 
(Originally:

Thanks for your comments on TEL-186 Zack! Would we need to do something like this in the base image first?

This PR is a conversation piece. I don't think it's mergeworthy yet.)

### CODE REVIEW

**Comment by zmoore-usgs** [comment]
This change wouldn't necessarily be _needed_ since a downstream image could define the `SERVER_PORT` ENV and override the Healthcheck to use that instead, but I'm certainly in favor of generalizing stuff in this image to make it more widely applicable (beyond just changing serverPort --> SERVER_PORT also changing all other wrapped Spring native envs to their native versions) and I think it's definitely a discussion worth having.

There's been some discussion about whether this base image is general enough as-is or if it needs to be made much more general in order to be widely applicable (and whether or not we start a new one from scratch or change this one has been debated as well), and create a much deeper image hierarchy as things get more specific, perhaps even going as far as a WMA-JAVA-BASE then --> WMA-SPRING-BOOT-BASE then --> DECAP(Or MLR, AQCU, etc..)-SPRING-BOOT-BASE --> service-specific images.

I think our base image strategy would be a good discussion topic for the week that Tim is in Middleton and we could consider using Telemetry as the basis of starting more proper base image hierarchy, should we choose to go that route.

**Comment by cschroed-usgs** [comment]
> This change wouldn't necessarily be _needed_ since a downstream image could define the `SERVER_PORT` ENV and override the Healthcheck to use that instead
Good point!

> but I'm certainly in favor of generalizing stuff in this image to make it more widely applicable (beyond just changing serverPort --> SERVER_PORT also changing all other wrapped Spring native envs to their native versions) and I think it's definitely a discussion worth having.

Cool! I will try updating the other variables too. I get that we might not want to merge those kinds of changes now pending discussion on base image re-organization, but maybe the variable switching will encourage some worthwhile discussion too?

**Comment by cschroed-usgs** [comment]
Wanted to add: I think it would be fun to experiment with a new image hierarchy in Telemetry. As of now, Telemetry is only planning on having one spring boot app deployed in a container. We're thinking of deploying some event-driven microservices via Lambda instead of into a containerized runtime. If that continues to be true, Telemetry might not be the best candidate for fleshing out the "project base -> microservice-specific " portion of the new hierarchy.

**Comment by cschroed-usgs** [comment]
I think I took all possible duplicate variables out of `application.yml`. Is this what you had in mind @zmoore-usgs ?

**Comment by zmoore-usgs** [comment]
Yeah, that's what I had in mind. One thing this did uncover that I forgot about was the situation of binding multiple Spring envs to a single system env. The application.yml included in this image is only for the included sample app and is removed in any downstream images that base off of this image. So services that build off of this base image will need to know that they either need to configure their own application.yml to have the `key-password` use the same ENV as the `key-store-password`, or know that they need to define the `key-store-password` ENV as well

**Comment by cschroed-usgs** [comment]
Hi Zack! Sorry this fell off the radar. I tried to address the issue you brought up by enabling users of the base image to give different passwords to a key and to a keystore. Does that resolve things?

**Comment by zmoore-usgs** [comment]
We may want to cut a release of this image as-is before merging this so if it does break stuff those projects can use the last release for now and then update things on their end to work with the newer version.

**Comment by cschroed-usgs** [comment]
Good ideas. I support releasing the image as-is to permit rollback. I would help do that, but I don't know how. Could you take a stab at it? Alternatively, if you're willing to teach a man to fish, I'm willing to learn :)

**Comment by zmoore-usgs** [comment]
The system we have setup is pretty easy, though a bit manual.

1. Cut a new release from the GitHub UI: https://github.com/USGS-CIDA/docker-wma-spring-boot-base/releases
(Not sure why our 2 more recent releases aren't showing in the releases tab.. They are showing in the Tags tab. I believe this next release would be 0.0.4)

2. In Jenkins run a build and select the new release tag as the Git Tag parameter in the build parameters

**Comment by zmoore-usgs** [comment]
Once a release is cut you may want to reach out to M&Ms to see what, if any, of there projects are using this image so that they're aware of the changes. If things break after this merge I can create tickets for MLR, AQCU, and Water Auth.

**Comment by cschroed-usgs** [comment]
I pushed some updates.

So it turns out that...
1. My test to verify that we could use different key and store passwords was bad; it didn't actually work.
2. most tools that use a PKCS12 keystore (including Java's keytool) require that the key password and the store password match.
3. If we don't set `SERVER_SSL_KEYPASSWORD`, Spring Boot's bundled Tomcat uses `SERVER_SSL_KEYSTOREPASSWORD` to unlock the keys

**Comment by cschroed-usgs** [comment]
Password mismatch error from `keytool`
```
wma-spring-boot-base | + keytool -v -importkeystore -deststorepass SOMETHING -destkeystore /home/spring/localkeystore.pkcs12 -deststoretype PKCS12 -srckeystore /home/spring/tomcat.pkcs12 -srcstorepass SOMETHING -srcstoretype PKCS12 -noprompt -destkeypass SOMETHING_ELSE
wma-spring-boot-base | Importing keystore /home/spring/tomcat.pkcs12 to /home/spring/localkeystore.pkcs12...
wma-spring-boot-base | Warning:  Different store and key passwords not supported for PKCS12 KeyStores. Ignoring user-specified -destkeypass value.
wma-spring-boot-base | keytool error: java.io.IOException: keystore password was incorrect
wma-spring-boot-base | java.io.IOException: keystore password was incorrect
wma-spring-boot-base |  at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2059)
wma-spring-boot-base |  at java.security.KeyStore.load(KeyStore.java:1445)
wma-spring-boot-base |  at sun.security.tools.keytool.Main.loadSourceKeyStore(Main.java:2040)
wma-spring-boot-base |  at sun.security.tools.keytool.Main.doCommands(Main.java:1067)
wma-spring-boot-base |  at sun.security.tools.keytool.Main.run(Main.java:366)
wma-spring-boot-base |  at sun.security.tools.keytool.Main.main(Main.java:359)
wma-spring-boot-base | Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
wma-spring-boot-base |  ... 6 more
wma-spring-boot-base exited with code 1
```

**Comment by zmoore-usgs** [comment]
Gotcha, that's good to know; I didn't know about either of those findings. So at this point I think this is good to go once the release is cut. The only other thing you may want to do is reach out to M&Ms to see if any of their projects are using the "latest" tag of this image and if so give them a heads-up about this merge and the 0.0.4 release. I can verify AQCU, MLR, and Water Auth on our end, but I'm not sure if anything on M&Ms is also using this image.

**Comment by cschroed-usgs** [comment]
Andrew reported that there are no usages of this base image on M+M projects.

**Comment by cschroed-usgs** [comment]
@zmoore-usgs thanks to you and Ivan for implementing versioned tags!  Would you feel comfortable about merging this now that we have the versioned tags?

File: secrets.env [hunk 1]
```diff
@@ -1,2 +1,2 @@
-keystorePassword=changeme
+SERVER_SSL_KEYSTOREPASSWORD=tomcat
```
Comment (zmoore-usgs) [inline secrets.env]: Should this be `changeme` ?

**Comment by zmoore-usgs** [comment]
Gave it one last quick overview and left one small question, but otherwise yeah I think this should be good to merge noe

**Pull request merged**